File: typescript/packages/mechanisms/evm/src/batch-settlement/storage/delegatedAuth.ts
import type { Network } from "@x402/core/types";

/**
 * Deposit-time caller identity bound to a channel so a later refund settle
 * can be correlated to the same service.
 */
export interface DelegatedAuthBinding {
  channelId: string;
  network: Network;
  /** Identity `resolveCallerIdentity` returned on the deposit settle. */
  callerIdentity: string;
}

/** Returned by {@link DelegatedAuthStore.bind} when a different identity owns the key. */
export class DelegatedAuthIdentityConflictError extends Error {
  /** Create an error when a channel already has a different delegated identity. */
  constructor() {
    super("delegated auth binding already exists for a different identity");
    this.name = "DelegatedAuthIdentityConflictError";
  }
}

/**
 * Pluggable store of delegated deposit/refund caller-identity bindings.
 *
 * `bind` is keyed by `(channelId, network)` and is first-writer-wins:
 *
 * - no existing row → insert
 * - existing row, same `callerIdentity` → success (idempotent retry)
 * - existing row, different `callerIdentity` → {@link DelegatedAuthIdentityConflictError}
 *
 * `get` returns `undefined` for not-found and propagates store errors so a
 * host can map infra failures separately from unauthenticated.
 */
export interface DelegatedAuthStore {
  bind(binding: DelegatedAuthBinding): Promise<void>;
  get(channelId: string, network: Network): Promise<DelegatedAuthBinding | undefined>;
  delete(channelId: string, network: Network): Promise<void>;
}

/**
 * In-memory {@link DelegatedAuthStore}. A multi-replica facilitator must
 * inject a shared implementation; a lost binding fails closed.
 */
export class InMemoryDelegatedAuthStore implements DelegatedAuthStore {
  private readonly bindings = new Map<string, DelegatedAuthBinding>();

  /**
   * Record the caller identity for a channel. First writer wins: a later
   * `bind` with the same identity is a no-op; a different identity is an error.
   *
   * @param binding - Channel, network, and identity
   */
  async bind(binding: DelegatedAuthBinding): Promise<void> {
    const key = bindingKey(binding.channelId, binding.network);
    const existing = this.bindings.get(key);
    if (existing) {
      if (existing.callerIdentity === binding.callerIdentity) {
        return;
      }
      throw new DelegatedAuthIdentityConflictError();
    }
    this.bindings.set(key, binding);
  }

  /**
   * Look up a binding.
   *
   * @param channelId - Channel id
   * @param network - CAIP-2 network the channel was opened on
   * @returns Stored binding, or undefined when absent
   */
  async get(channelId: string, network: Network): Promise<DelegatedAuthBinding | undefined> {
    const binding = this.bindings.get(bindingKey(channelId, network));
    return binding ? { ...binding } : undefined;
  }

  /**
   * Remove a binding.
   *
   * @param channelId - Channel id
   * @param network - CAIP-2 network the channel was opened on
   */
  async delete(channelId: string, network: Network): Promise<void> {
    this.bindings.delete(bindingKey(channelId, network));
  }
}

/**
 * Composite key so the same channel id on two networks cannot collide.
 *
 * @param channelId - Channel id
 * @param network - CAIP-2 network
 * @returns Store key
 */
function bindingKey(channelId: string, network: Network): string {
  return `${network}:${channelId.toLowerCase()}`;
}

File: go/mechanisms/evm/batch-settlement/storage/delegated_auth.go
package storage

import (
	"context"
	"strings"
	"sync"
)

// DelegatedAuthBinding is a deposit-time caller identity bound to a channel
// so a later refund settle can be correlated to the same service.
type DelegatedAuthBinding struct {
	ChannelId      string
	Network        string
	CallerIdentity string
}

// DelegatedAuthIdentityConflictError is returned by Bind when a different
// identity already owns the (channelId, network) key.
type DelegatedAuthIdentityConflictError struct{}

func (e *DelegatedAuthIdentityConflictError) Error() string {
	return "delegated auth binding already exists for a different identity"
}

// DelegatedAuthStore persists delegated deposit/refund caller-identity bindings.
// Bind is durable; nil means a later Get sees the row. Do not broadcast before that.
type DelegatedAuthStore interface {
	// Bind is first-writer-wins. inserted is true only when this call created the row.
	// Same identity returns inserted == false; a different identity conflicts.
	Bind(ctx context.Context, binding DelegatedAuthBinding) (inserted bool, err error)
	Get(ctx context.Context, channelId string, network string) (*DelegatedAuthBinding, error)
	Delete(ctx context.Context, channelId string, network string) error
}

// InMemoryDelegatedAuthStore is a volatile DelegatedAuthStore. A multi-replica
// facilitator must inject a shared implementation; a lost binding fails closed.
type InMemoryDelegatedAuthStore struct {
	mu       sync.Mutex
	bindings map[string]DelegatedAuthBinding
}

var _ DelegatedAuthStore = (*InMemoryDelegatedAuthStore)(nil)

// NewInMemoryDelegatedAuthStore creates an empty in-memory binding store.
func NewInMemoryDelegatedAuthStore() *InMemoryDelegatedAuthStore {
	return &InMemoryDelegatedAuthStore{bindings: make(map[string]DelegatedAuthBinding)}
}

// Bind records the caller identity. inserted is true only for a new row.
func (s *InMemoryDelegatedAuthStore) Bind(_ context.Context, binding DelegatedAuthBinding) (bool, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	key := bindingKey(binding.ChannelId, binding.Network)
	existing, ok := s.bindings[key]
	if ok {
		if existing.CallerIdentity == binding.CallerIdentity {
			return false, nil
		}
		return false, &DelegatedAuthIdentityConflictError{}
	}
	s.bindings[key] = binding
	return true, nil
}

// Get looks up a binding. The returned value is a copy so callers cannot
// mutate the stored row.
func (s *InMemoryDelegatedAuthStore) Get(_ context.Context, channelId string, network string) (*DelegatedAuthBinding, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	binding, ok := s.bindings[bindingKey(channelId, network)]
	if !ok {
		return nil, nil
	}
	cp := binding
	return &cp, nil
}

// Delete removes a binding.
func (s *InMemoryDelegatedAuthStore) Delete(_ context.Context, channelId string, network string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.bindings, bindingKey(channelId, network))
	return nil
}

func bindingKey(channelId string, network string) string {
	return network + ":" + strings.ToLower(channelId)
}
